controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +124,296 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+	pvReclaimPolicy := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.storageClassClaim.Name,
+			Namespace: r.storageClassClaim.Namespace,
+			Annotations: map[string]string{
+				"description": "Provides RWO and RWX Filesystem volumes",
+			},
+			Finalizers: []string{
+				StorageClassFinalizer,
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.cephfs.csi.ceph.com", r.storageCluster.Namespace),
+		Parameters:           data,
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) getCephRBDStorageClass(data map[string]string) *storagev1.StorageClass {
+	pvReclaimPolicy := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.storageClassClaim.Name,
+			Namespace: r.storageClassClaim.Namespace,
+			Annotations: map[string]string{
+				"description": "Provides RWO Filesystem volumes, and RWO and RWX Block volumes",
+			},
+			Finalizers: []string{
+				StorageClassFinalizer,
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.rbd.csi.ceph.com", r.storageCluster.Namespace),
+		Parameters:           data,
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) createOrReplaceStorageClass(storageClass *storagev1.StorageClass) error {
+	existing := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: r.storageClassClaim.Name,
+		},
+	}
+	if err := r.get(existing); err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// Already exists, need to delete it first.
+	if existing.UID != "" {
+
+		// Since we have to update the existing StorageClass, so we will delete the existing storageclass and create a new one.
+		r.log.Info("StorageClass needs to be updated, deleting it.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+
+		// Check if StorageClass is in use, if yes, then fail.
+		// Wait until all PVs using the StorageClass under deletion are removed.
+		// Check for any PVs using the StorageClass.
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
+		}
+
+		// Remove finalizer from StorageClass.
+		storageClass.SetFinalizers(remove(storageClass.Finalizers, StorageClassFinalizer))
+		if err := r.update(storageClass); err != nil {
+			return fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+		}
```

Comment:
you cannot set finalizers on the storageclass it's a cluster scoped resource.

Reply:
Removing.

Reply:
sorry discard this one.